### DISCUSSION TITLE
Merge PDF support

### DISCUSSION THREAD

**Discussion by robtomlinson** [QuestPDF/QuestPDF#192] [Ideas]
I am unfamiliar with the PDF specification, but I would like to have the ability to merge existing PDFs into a QuestPDF document. This is a feature of many other commercially available C# PDF libraries, but I don't know if there are any proprietary licensing, etc. that may be required that may prevent this from being added to QuestPDF.
Are there many technical/licensing hurdles that would prevent such an addition?

**Comment by MarcinZiabek** [comment]
Hello 😁 Currently, due to architecture limitations, the library focuses only on PDF file generation. QuestPDF acts as a layout engine and uses SkiaSharp to produce the final PDF file. Therefore, it does not have any PDF content layer that could be used to implement altering and merging functionalities. It is just too big scope to handle at once 😁

**Comment by skrabbe** [comment]
Hi Marcin,

Is there any changes for this question in any near future or you dont think it will ever happen?
All of our customer are using our software for printing invoices and many of these need some kind of attachment, often in another pdf-document.

**Reply by MarcinZiabek** [reply]
Right now, I am slowly finishing the introduction of a [dual-license model](https://www.questpdf.com/pricing.html). This will hopefully provide some extra funding and ensure my availability to implement this and many other features 😁 

Till then, I recommend to continue using QuestPDF for document generation, and utilizing [MigraDoc](http://www.pdfsharp.net/Features.ashx) for the merge functionality.

**Comment by DawidRobertNowak** [comment]
Hello @skrabbe,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

The functionality you described is now available through the `DocumentOperation` API, introduced in version 2024.12.0. You can load and merge existing PDF files natively. Full documentation is available here:
https://www.questpdf.com/concepts/document-operations.html

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
